micromanager_gui/_core_widgets/_presets_widget.py

```diff
@@ -0,0 +1,139 @@
+
+        self._mmc.events.configSet.connect(self._on_cfg_set)
+        self._mmc.events.configSet.connect(self._highlight_if_preset_state_changed)
+        self._mmc.events.systemConfigurationLoaded.connect(self.refresh)
+        self._mmc.events.propertyChanged.connect(self._highlight_if_prop_changed)
+
+        self.destroyed.connect(self.disconnect)
+
+    def _on_text_activate(self, text: str):
+        """Used in case there is only one preset"""
+        if len(self._presets) == 1:
+            self._mmc.setConfig(self._group, text)
+            set_wdg_color(self.text_color, self._combo)
+
+    def _on_combo_changed(self, text: str) -> None:
+        self._mmc.setConfig(self._group, text)
+        set_wdg_color(self.text_color, self._combo)
+
+    def _on_cfg_set(self, group: str, preset: str) -> None:
+        if group == self._group and self._combo.currentText() != preset:
+            with signals_blocked(self._combo):
+                self._combo.setCurrentText(preset)
+                set_wdg_color(self.text_color, self._combo)
+
+    def value(self) -> str:
+        return self._combo.currentText()
+
+    def setValue(self, value: str) -> None:
+        if value not in self._mmc.getAvailableConfigs(self._group):
+            raise ValueError(
+                f"{value!r} must be one of {self._mmc.getAvailableConfigs(self._group)}"
+            )
+        self._combo.setCurrentText(str(value))
+
+    def allowedValues(self) -> Tuple[str]:
+        return tuple(self._combo.itemText(i) for i in range(self._combo.count()))
+
+    def refresh(self) -> None:
+        with signals_blocked(self._combo):
+            self._combo.clear()
+            if self._group not in self._mmc.getAvailableConfigGroups():
+                self._combo.addItem(f"No group named {self._group}.")
+                self._combo.setEnabled(False)
+            else:
+                presets = self._mmc.getAvailableConfigs(self._group)
+                self._combo.addItems(presets)
+                self._combo.setEnabled(True)
+
+    def _highlight_if_preset_state_changed(self, group: str, preset: str):
+        """Set the text color to magenta if a preset has changed"""
+
+        if group == self._group:
+            return
+
+        dp = get_dev_prop(group, preset)
+        if any(x for x in dp if x in self.dev_prop):
+            self._set_if_props_match_preset()
+
+    def _highlight_if_prop_changed(self, device: str, property: str, value: str):
+        """Set the text color to magenta if a preset property has changed"""
+
+        if (device, property) not in self.dev_prop:
+            if self._mmc.getDeviceType(device) != DeviceType.StateDevice:
+                return
+            # a StateDevice has also a "Label" property. If "Label" is not
+            # in dev_prop, we check if the property "State" is in dev_prop.
+            if (device, "State") not in self.dev_prop:
+                return
+
+        self._set_if_props_match_preset()
```

Comment:
I think you can combine this an `_set_if_props...` into one function, because all this does is a little prep work and then calls the other function.

Reply:
@ianhi I was thinking that we need to keep the part below only for the `propertyChanged` signal. The `configSet` signal emits `group` and `preset` and not (`device`, `property`, `value`). I wanted to run the `_set_if_props_match_preset` method only if the (device, property) is included in the preset. What do you think?
```
`if (device, property) not in self.dev_prop:
            if self._mmc.getDeviceType(device) != DeviceType.StateDevice:
                return
            # a StateDevice has also a "Label" property. If "Label" is not
            # in dev_prop, we check if the property "State" is in dev_prop.
            if (device, "State") not in self.dev_prop:
                return
```